### PR TITLE
[KYUUBI #2289] Use unique tag to kill applications 

### PR DESCRIPTION
**Pull request by Nick-0723**
### _Why are the changes needed?_

Use unique tag to kill applications instaed of log
#2289 

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 1]
```diff
@@ -193,28 +194,32 @@ class SparkProcessBuilder(
...
+          s"Failed to kill Application $applicationId tagged with $engineRefId," +
+            s" please kill it manually. Caused by ${e.getMessage}."
+      } finally {
+        if (yarnClient != null) {
+          yarnClient.stop()
         }
-      case None => ""
+      }
+    } else {
+      ""
```
Comment (zwangsheng) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:220]: How about we add a warning here to tell that killApplication currently only works with YARN ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 2]
```diff
@@ -193,28 +194,32 @@ class SparkProcessBuilder(
     }
   }
 
-  override def killApplication(line: String = lastRowsOfLog.toArray.mkString("\n")): String =
-    YARN_APP_NAME_REGEX.findFirstIn(line) match {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:197]: YARN_APP_NAME_REGEX is not used anymore

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderSuite.scala [hunk 3]
```diff
@@ -240,21 +240,22 @@ class SparkProcessBuilderSuite extends KerberizedTestHelper with MockitoSugar {
...
     val pb1 = new FakeSparkProcessBuilder(conf) {
       override protected def env: Map[String, String] = Map()
       override def getYarnClient: YarnClient = mock[YarnClient]
     }
-    val exit1 = pb1.killApplication("21/09/30 17:12:47 INFO yarn.Client: " +
-      "Application report for application_1593587619692_20149 (state: ACCEPTED)")
-    assert(exit1.contains("Killed Application application_1593587619692_20149 successfully."))
+    pb1.conf.set("spark.master", "yarn")
+    val exit1 = pb1.killApplication(engineRefId)
+    assert(exit1.equals(s"There are no Application tagged with $engineRefId"))
```
Comment (zwangsheng) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderSuite.scala:251]: If possible, add a successful kill application case here.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala [hunk 4]
```diff
@@ -185,6 +186,12 @@ trait ProcBuilder {
     process
   }
 
+  def killApplicationWrap(engineRefId: String): String =
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala:189]: I have an alternative idea for it, change the method signature to `def killApplication(clue: Either[String, String]): String`, WDYT @Nick-0723 @yaooqinn?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 5]
```diff
@@ -193,28 +194,32 @@ class SparkProcessBuilder(
...
-        } catch {
-          case e: Throwable =>
-            s"Failed to kill Application $appId, please kill it manually." +
-              s" Caused by ${e.getMessage}."
-        } finally {
-          if (yarnClient != null) {
-            yarnClient.stop()
-          }
+  override def killApplication(engineRefId: String): String = {
+    val master = conf.getOption("spark.master").getOrElse("local")
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:198]: nit: don't need to be `local`

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 6]
```diff
@@ -193,28 +194,32 @@ class SparkProcessBuilder(
...
+        if (apps.isEmpty) return s"There are no Application tagged with $engineRefId"
+        applicationId = apps.asScala.head.getApplicationId
+        yarnClient.killApplication(applicationId)
+        s"Killed Application $applicationId successfully."
+      } catch {
+        case e: Throwable =>
+          s"Failed to kill Application $applicationId tagged with $engineRefId," +
+            s" please kill it manually. Caused by ${e.getMessage}."
+      } finally {
+        if (yarnClient != null) {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:215]: yarn client nevel null

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 7]
```diff
@@ -193,28 +194,32 @@ class SparkProcessBuilder(
...
+    if (master.equals("yarn")) {
+      var applicationId: ApplicationId = null
+      try {
+        val hadoopConf = KyuubiHadoopUtils.newHadoopConf(conf)
+        yarnClient.init(hadoopConf)
+        yarnClient.start()
+        val apps = yarnClient.getApplications(null, null, Set(engineRefId).asJava)
+        if (apps.isEmpty) return s"There are no Application tagged with $engineRefId"
+        applicationId = apps.asScala.head.getApplicationId
+        yarnClient.killApplication(applicationId)
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:208]: `yarnClient.killApplication(applicationId, clue)` 
it's better to add a kyuubi-specific clue head for the kill reason

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala (hunk 4 above)
Comment (Nick-0723) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala:189]: So we judge whether it is Left or Right in EngineRef? We can't get logs in EngineRef

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderSuite.scala (hunk 3 above)
Comment (Nick-0723) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderSuite.scala:251]: Actually I don't know how to add yarn test cases

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala (hunk 4 above)
Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala:189]: Yeah, I mean merge two methods to one. And use `Left` to represent  `engineRefId` and Right to represent `line`

Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala:189]: the right will be removed eventually

Comment (Nick-0723) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala:189]: > Yeah, I mean merge two methods to one. And use `Left` to represent `engineRefId` and Right to represent `line`

If it is merged into one, the implementation of the flink engine will also needs to be modified

Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala:189]: I think we don't need such thing, the tag-based killing is promising

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderSuite.scala (hunk 3 above)
Comment (turboFei) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderSuite.scala:251]: FYI: MiniYarnService

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala (hunk 5 above)
Comment (cxzl25) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:198]: In this way, `spark.master` can only be configured in `kyuubi-defaults.conf` to support automatic kill yarn app, but not in `spark-defaults.conf`?

nit:`spark.master` can be used as a constant of `object SparkProcessBuilder`.

Comment (Nick-0723) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:198]: Yes, we can't get the configuration in spark-defaults.conf, or is there a way I don't know?
cc @yaooqinn

Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:198]: We can get spark.master conf here; or else from spark-defaults.conf from "SPARK_CONF_DIR"; or else SPARK_HOME/conf

**Comment by ulysses-you** [comment]
there are some conflicts

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala (hunk 5 above)
Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:198]: FYI:
```
  private def getSparkDefaultsConf(): Map[String, String] = {
    val sparkDefaultsConfFile = env.get("SPARK_CONF_DIR")
      .orElse(env.get("SPARK_HOME").map(_ + "/conf"))
      .map(_ + "/spark-defaults.conf").map(new File(_))
      .filter(_.exists())
    Utils.getPropertiesFromFile(sparkDefaultsConfFile)
  }

conf.getOption("spark.master").getOrElse(getSparkDefaultsConf().get("spark.master"))
```

Comment (Nick-0723) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:198]: thanks， I saw your pr is drafted，I’ll add it to this pr

**Comment by Nick-0723** [comment]
seems flaky test failures, can you help? @yaooqinn 

```
- hive process builder *** FAILED ***
  org.apache.kyuubi.KyuubiSQLException: JAVA_HOME is not set! For more information on installing and configuring JAVA_HOME, please visit https://kyuubi.apache.org/docs/latest/deployment/settings.html#environments
  at org.apache.kyuubi.KyuubiSQLException$.apply(KyuubiSQLException.scala:69)
  at org.apache.kyuubi.engine.ProcBuilder.validateEnv(ProcBuilder.scala:326)
  at org.apache.kyuubi.engine.ProcBuilder.validateEnv$(ProcBuilder.scala:325)
  at org.apache.kyuubi.engine.hive.HiveProcessBuilder.validateEnv(HiveProcessBuilder.scala:33)
  at org.apache.kyuubi.engine.ProcBuilder.executable(ProcBuilder.scala:52)
  at org.apache.kyuubi.engine.ProcBuilder.executable$(ProcBuilder.scala:49)
  at org.apache.kyuubi.engine.hive.HiveProcessBuilder.executable(HiveProcessBuilder.scala:33)
  at org.apache.kyuubi.engine.hive.HiveProcessBuilder.commands(HiveProcessBuilder.scala:47)
  at org.apache.kyuubi.engine.hive.HiveProcessBuilder.toString(HiveProcessBuilder.scala:89)
  at org.apache.kyuubi.engine.hive.HiveProcessBuilderSuite.$anonfun$new$1(HiveProcessBuilderSuite.scala:29)
```

**Comment by yaooqinn** [comment]
> HiveProcessBuilderSuite

Can you make a follow up get getting JAVA_HOME with falling back to `System.getProperty("java.home")`

**Comment by Nick-0723** [comment]
> > HiveProcessBuilderSuite
> 
> Can you make a follow up get getting JAVA_HOME with falling back to `System.getProperty("java.home")`

I can't reproduce this error locally

image

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala [hunk 8]
```diff
@@ -78,7 +78,12 @@ class FlinkProcessBuilder(
 
   override protected def commands: Array[String] = Array(executable)
 
-  override def killApplication(line: String = lastRowsOfLog.toArray.mkString("\n")): String = {
+  override def killApplication(clue: Either[String, String]): String = clue match {
+    case Left(_) => ""
```
Comment (turboFei) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala:82]: is it difficult to add engine tag for Flink engine?

Comment (Nick-0723) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala:82]: Kent said to deal with Spark first, should Flink be put in another pr?

Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala:82]: ok， we can do that in another pr

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderOnYarnSuite.scala [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  test("test kill application") {
+    val engineRefId = UUID.randomUUID().toString
+
+    conf.set(
+      SparkProcessBuilder.TAG_KEY,
+      conf.getOption(SparkProcessBuilder.TAG_KEY).map(_ + ",").getOrElse("") +
+        "KYUUBI," + engineRefId)
+    val builder = new SparkProcessBuilder(Utils.currentUser, conf)
+    val proc = builder.start
+    proc.waitFor(30, TimeUnit.SECONDS)
```
Comment (turboFei) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderOnYarnSuite.scala:46]: nit: is `eventually` better?

**Comment by turboFei** [comment]
left some comment, LGTM overall

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderOnYarnSuite.scala (hunk 9 above)
Comment (Nick-0723) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderOnYarnSuite.scala:46]: `yarnClient` closes after `killApplication` is called once, I tried `eventually`

Comment (turboFei) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderOnYarnSuite.scala:46]: I saw that
```
  val yarnClient = getYarnClient
```
is only used in method `killApplicationByTag`.

So, I think you can make `yarnClient` as an temporary variable in `killApplicationByTag`.

Comment (Nick-0723) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderOnYarnSuite.scala:46]: I’m not sure we need to use singleton here?

Comment (turboFei) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderOnYarnSuite.scala:46]: We do not need to use singleton I think.

Comment (Nick-0723) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderOnYarnSuite.scala:46]: ok, tks

**Review by turboFei — APPROVED** [review]

**Comment by Nick-0723** [comment]
If you have time, please help review this. @yaooqinn

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala [hunk 10]
```diff
@@ -78,7 +78,12 @@ class FlinkProcessBuilder(
 
   override protected def commands: Array[String] = Array(executable)
 
-  override def killApplication(line: String = lastRowsOfLog.toArray.mkString("\n")): String = {
+  override def killApplication(clue: Either[String, String]): String = clue match {
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala:81]: Why do we need this change ? For Spark SQL engine, we can call `killApplication(engineRefId)` and others call `killApplication()`

Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala:81]: And override the `killApplication` in `SparkProcessBuilder`.  BTW, we can rename the input parameter of `killApplication` to engineRefId

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala (hunk 4 above)
Comment (Nick-0723) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala]: @ulysses-you , I think this conversation will solve your question

Comment (Nick-0723) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala]: @ulysses-you , I think this conversation will solve your question

**Review by pan3793 — APPROVED** [review]

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala (hunk 4 above)
Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ProcBuilder.scala]: I have read this history.. just think it's unnecessary

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala (hunk 10 above)
Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala:81]: > ... and others call `killApplication()`

Then it is not testable

Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala:81]: > ... and others call `killApplication()`

Then it is not testable

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**